src/Core/Framework/App/Api/ShopIdController.php

```diff
@@ -0,0 +1,95 @@
+use Shopware\Core\Framework\DataAbstractionLayer\Search\Criteria;
+use Shopware\Core\Framework\DataAbstractionLayer\Search\Filter\NotEqualsFilter;
+use Shopware\Core\Framework\Log\Package;
+use Shopware\Core\Framework\Routing\ApiRouteScope;
+use Shopware\Core\PlatformRequest;
+use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;
+use Symfony\Component\HttpFoundation\JsonResponse;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\Routing\Attribute\Route;
+
+/**
+ * @internal
+ */
+#[Route(defaults: [PlatformRequest::ATTRIBUTE_ROUTE_SCOPE => [ApiRouteScope::ID]])]
+#[Package('framework')]
+class ShopIdController extends AbstractController
+{
+    /**
+     * @param EntityRepository<AppCollection> $appRepository
+     */
+    public function __construct(
+        private readonly Resolver $shopIdChangeResolver,
+        private readonly ShopIdProvider $shopIdProvider,
+        private readonly EntityRepository $appRepository,
+    ) {
+    }
+
+    #[Route(path: 'api/app-system/shop-id/change-strategies', name: 'api.app_system.shop_id.change_strategies', methods: ['GET'])]
+    public function getAvailableStrategies(): JsonResponse
+    {
+        return new JsonResponse($this->shopIdChangeResolver->getAvailableStrategies());
+    }
+
+    #[Route(path: 'api/app-system/shop-id/change', name: 'api.app_system.shop_id.change', methods: ['POST'])]
+    public function changeShopId(Request $request, Context $context): Response
+    {
+        $strategy = $request->get('strategy');
+
+        if (!$strategy) {
+            throw AppException::missingRequestParameter('strategy');
+        }
+
+        try {
+            $this->shopIdChangeResolver->resolve($strategy, $context);
+        } catch (ShopIdChangeStrategyNotFoundException $e) {
+            throw AppException::shopIdChangeResolveStrategyNotFound($strategy);
+        }
+
+        return new Response(null, Response::HTTP_NO_CONTENT);
+    }
+
+    #[Route(path: 'api/app-system/shop-id/check', name: 'api.app_system.shop_id.check', methods: ['POST'])]
+    public function checkShopId(): Response
+    {
+        try {
+            $this->shopIdProvider->getShopId();
+        } catch (ShopIdChangeSuggestedException $e) {
+            return new JsonResponse([
+                'apps' => $this->appsRegisteredAtAppServers(),
+                'fingerprints' => $e->comparisonResult,
+            ]);
+        }
+
+        return new Response(null, Response::HTTP_NO_CONTENT);
+    }
+
+    /**
+     * @return list<string>
+     */
+    private function appsRegisteredAtAppServers(): array
+    {
+        $criteria = new Criteria();
+        $criteria->addFilter(new NotEqualsFilter('appSecret', null));
+
+        $apps = $this->appRepository->search($criteria, Context::createDefaultContext())->getElements();
+
+        return array_values(array_map(function (AppEntity $app) {
```

Comment:
if you do not call `getElements` but `getEntities` instead, you can also make use of `\Shopware\Core\Framework\Struct\Collection::map`